### PR TITLE
Pipe Amending

### PR DESCRIPTION
**Pull request by Sonnicon**
## About The Pull Request

When one wants to connect to an existing pipe, the only ways to do so are by either having previously planned and added a free node, or removing a pipe and adding a free node. This is a problem when in a multiplayer environment as planning is usually not possible, and removing a pipe will often result in a gas leak. This merge request adds a way to add nodes to existing pipes, without unwrenching the pipe and causing trouble. This is done by rightclicking a pipe with an RPD that has the Amend upgrade, and selecting a direction for a new node from a radial menu. Also works on heat exchange pipes. The upgrade must be researched from the Rapid Construction node, and can be printed in the engineering lathe. Previous RPD rightclick functionality (copying layer and colour of existing pipe) has been replaced (the colour part didn't work anyway).
Additionally, placing multiple connected pipes in a single atmos tick will no longer cause a runtime error.

## Why It's Good For The Game

Makes piping less annoying, especially with multiple people.
Less errors is good.

## Changelog
:cl:
add: RPD Amend upgrade disk, allows an RPD to rightclick atmos pipes to add nodes.
fix: Placing pipes quickly won't runtime.
/:cl:

### CODE REVIEW

**Comment by rdragan** [comment]
I'm not really a fan of this. I think there should be consequences for messing up your setups or failing to plan adequately. I also don't understand what you mean by planning not being usually possible - you can absolutely plan ahead every time; sometimes the plans will get interrupted but that's part of the game and having to come with solutions on the spot is what makes this fun.  

And it's not impossible to plug into an existing pipenet already, it just takes more care and knowledge of how the pipe system works. This is part of the skill set of an experienced atmos tech. Leaks only occur when the atmos tech doesn't know what they're doing. The plasma leak and ensuing fire is a learning experience, I don't want atmos baby-proofed.

File: code/modules/atmospherics/machinery/pipes/pipes.dm [hunk 1]
```diff
@@ -8,12 +10,23 @@
...
 	can_buckle = TRUE
 	buckle_requires_restraints = TRUE
 	buckle_lying = 90
 
+	// Lists can't have integer keys
+	var/static/list/radial_options = list(
+		"NORTH" = ARROW_ICON(NORTH),
+		"EAST" = ARROW_ICON(EAST),
+		"SOUTH" = ARROW_ICON(SOUTH),
+		"WEST" = ARROW_ICON(WEST)
```
Comment (TheChosenEvilOne) [inline code/modules/atmospherics/machinery/pipes/pipes.dm:25]: You could do "[NORTH]" "[EAST]" etc. and use text2num

File: code/modules/atmospherics/machinery/pipes/pipes.dm [hunk 2]
```diff
@@ -68,6 +81,73 @@
...
+			if("EAST")
+				direction = EAST
+			if("SOUTH")
+				direction = SOUTH
+			if("WEST")
+				direction = WEST
+
+		// Don't be already connected there
+		if(GetInitDirections() & direction)
+			to_chat(user, "<span class='warning'>There is already a connection in that direction!</span>")
```
Comment (TheChosenEvilOne) [inline code/modules/atmospherics/machinery/pipes/pipes.dm:104]:
```suggestion
			to_chat(user, span_warning("There is already a connection in that direction!"))
```

File: code/modules/atmospherics/machinery/pipes/pipes.dm [hunk 3]
```diff
@@ -68,6 +81,73 @@
...
+			return FALSE
+		// Don't overlap other pipes
+		for(var/obj/machinery/atmospherics/other in loc)
+			if((other.piping_layer != piping_layer) && !((other.pipe_flags | pipe_flags) & PIPING_ALL_LAYER)) // Don't continue if either pipe goes across all layers
+				continue
+			if(other.GetInitDirections() & direction) // New connection is occupied by other
+				to_chat(user, "<span class='warning'>There is already a pipe at that location!</span>")
```
Comment (TheChosenEvilOne) [inline code/modules/atmospherics/machinery/pipes/pipes.dm:111]:
```suggestion
				to_chat(user, span_warning("There is already a pipe at that location!"))
```

File: code/modules/atmospherics/machinery/pipes/heat_exchange/manifold.dm [hunk 4]
```diff
@@ -47,3 +48,6 @@
 /obj/machinery/atmospherics/pipe/heat_exchanging/manifold/layer4
 	piping_layer = 4
 	icon_state = "manifold-4"
+
+/obj/machinery/atmospherics/pipe/heat_exchanging/manifold/createAmend(turf/T, direction)
+	return new /obj/machinery/atmospherics/pipe/heat_exchanging/manifold4w/(T)
```
Comment (TheChosenEvilOne) [inline code/modules/atmospherics/machinery/pipes/heat_exchange/manifold.dm:53]:
```suggestion
	return new /obj/machinery/atmospherics/pipe/heat_exchanging/manifold4w(T) 
```

**Comment by Sonnicon** [comment]
There are still plenty of consequences for messing up setups - I intentionally didn't make it possible to remove pipes, so if you connect the wrong pipes together or otherwise MAKE something wrong, you will need to unwrench. This PR alleviates far planning, for example when you finished everything you wanted to do and decide to make another gas - you could take advantage of your existing infrastructure more easily. Planning isn't always possible in multiplayer scenarios, where you don't know who the others are prior. For example someone set up the turbine, and you want to create a burn canister. Easiest way currently is to build completely new pipes from the source tanks (unless nobody installed an adapter, but that's a serious planning failure), but with this PR you can make use of the other's pipes. Collaboration in atmos isn't all that common, so I think this is a big improvement.
You say that coming up with solutions on the spot is fun, but the problems that arise from unwrenching pipes to add a component boil down to _placing a holofan, bringing an air scrubber, and WAITING_. This requires no creativity from the player, and is unfun. My change introduces more unique problems that you can encounter and have to solve that involve more interaction with the atmospherics system - for example, you (OR SOMEONE ELSE) could connect to a pure pipe directly and contaminate it with another gas. This wouldn't previously happen because it was easier to build completely new pipes, rather than connect to existing ones. Such a problem requires identifying contaminated pipes, isolating the issue, and designing a filtering system to rememdy it, and is therefore much more involved than current roadblocks (waiting). Effective planning from the perpetrator would have not led to this problem, so planning is still required.

Planning a setup prior to construction still benefits you, but it shouldn't be something that is absolutely required every time; rounds will be short, and spending half of the time on planning isn't fun for most, and those that enjoy it still can.
Dangerous leaks will still happen, I doubt that someone who would cause a plasma leak would be able to effectively make use of this without placing a bad pipe they'd have to remove (cause a leak) later.

File: code/modules/atmospherics/machinery/pipes/simple.dm [hunk 5]
```diff
@@ -30,3 +31,8 @@
 /obj/machinery/atmospherics/pipe/simple/update_icon_state()
 	icon_state = "pipe[nodes[1] ? "1" : "0"][nodes[2] ? "1" : "0"]-[piping_layer]"
 	return ..()
+
+/obj/machinery/atmospherics/pipe/simple/createAmend(turf/T, direction)
+	var/obj/machinery/atmospherics/pipe/manifold/new_pipe = new /obj/machinery/atmospherics/pipe/manifold/(T)
+	new_pipe.setDir((initialize_directions | direction) ^ 15)
+	return new_pipe
```
[the remark this answers is not in the document]
Comment (Sonnicon) [reply code/modules/atmospherics/machinery/pipes/simple.dm:38]: Yeah this change doesn't work, it runtimes.

**Comment by Sparkezel** [comment]
I'd ask you to please break up your pr into your feature and the fix for pipe runtimes.

I sort of do like what you did, but I also fear what it might promote for the future atmos players. You could still hold conversation with other atmosians to ask them for safe outlets/intlets from their pipe masterpieces using layer adapters or manifolds.

And when it comes to designing pipe nets, NEVER start from the source, work from your output to your input to keep your pipe net clean exactly so you can redesign it on the spot without separating the pipes out. I think it would work out better if it was a separate, maybe researchable device that just cut out that one pipe so you can place in a new one.

**Comment by Sonnicon** [comment]
I have migrated this functionality from pipes into the RPD, gated behind a new RPD upgrade that can be researched.

**Review by Canohaaaaaaa — APPROVED** [review]

**Pull request merged**